### PR TITLE
Support deleting cloud accounts

### PR DESCRIPTION
**Pull request by gnmahanth**
Changes proposed in this pull request:
- https://github.com/deepfence/ThreatMapper/issues/2044
-
-

### CODE REVIEW

**Review by ramanan-ravi — APPROVED** [review]

File: deepfence_worker/tasks/scans/delete_cloud_accounts.go [hunk 1]
```diff
@@ -0,0 +1,283 @@
...
+		log.Error().Err(err).Msgf("failed to delete cloud resources for account %s", accountID)
+		return err
+	}
+
+	// delete hosts discovered from cloud
+	deleteHostsQuery := `
+	MATCH (n:Node{account_id: $node_id})
+	WHERE n.agent_running=false
+	DETACH DELETE n
+	`
```
Comment (noboruma) [inline deepfence_worker/tasks/scans/delete_cloud_accounts.go:171]: I think we should not delete the `Node`  directly and leave this job to the clean up task, there are couple of edge cases to be handled
Let's mark it as not running via:
```
MATCH (n:Node{account_id: $node_id})
SET n.agent_running=false
```
Instead

File: deepfence_server/handler/cloud_node.go [hunk 2]
```diff
@@ -324,3 +325,42 @@ func (h *Handler) CachePostureProviders(ctx context.Context) error {
...
+
+		data, err := json.Marshal(req)
+		if err != nil {
+			log.Error().Err(err).Msg("failed to marshal cloud account delete request")
+			h.respondError(&InternalServerError{err}, w)
+			return
+		}
+
+		if err := worker.Enqueue(utils.DeleteCloudAccounts, data, utils.CritialTaskOpts()...); err != nil {
+			log.Error().Err(err).Msg("failed enqueue task ddelete cloud accounts")
```
Comment (ibreakthecloud) [inline deepfence_server/handler/cloud_node.go:357]:
```suggestion
			log.Error().Err(err).Msg("failed enqueue task delete cloud accounts")
```

File: deepfence_server/handler/cloud_node.go [hunk 3]
```diff
@@ -324,3 +325,42 @@ func (h *Handler) CachePostureProviders(ctx context.Context) error {
...
+			h.respondError(&InternalServerError{err}, w)
+			return
+		}
+	}
+
+	h.AuditUserActivity(r, EventComplianceScan, ActionDelete, req, true)
```
Comment (ibreakthecloud) [inline deepfence_server/handler/cloud_node.go:363]: I think auditlogs should be created even if the task fails. 
we need a better framework at router level for auditlog, instrumenting all the methods with `h.AuditUserActivity` is overkill.(not scope of this PR)

File: deepfence_worker/tasks/scans/delete_cloud_accounts.go [hunk 4]
```diff
@@ -0,0 +1,283 @@
...
+
+	if err := json.Unmarshal(task.Payload(), &req); err != nil {
+		log.Error().Err(err).Msg("failed to decode cloud account delete request")
+		return err
+	}
+
+	log.Info().Msgf("delete cloud accounts payload: %v", req)
+
+	// delete accounts
+	for _, accontID := range req.NodeIDs {
```
Comment (ibreakthecloud) [inline deepfence_worker/tasks/scans/delete_cloud_accounts.go:35]:
```suggestion
	for _, accountID := range req.NodeIDs {
```

File: deepfence_worker/tasks/scans/delete_cloud_accounts.go [hunk 5]
```diff
@@ -0,0 +1,283 @@
...
+
+	log.Info().Msgf("delete cloud account %s", accountID)
+
+	org, err := isOrgAccount(ctx, accountID)
+	if err != nil {
+		log.Error().Err(err).Msgf("failed to determine if org account %s", accountID)
+		return err
+	}
+	if org {
+		childern, err := listOrgChildAccounts(ctx, accountID)
```
Comment (ibreakthecloud) [inline deepfence_worker/tasks/scans/delete_cloud_accounts.go:64]:
```suggestion
		children, err := listOrgChildAccounts(ctx, accountID)
```

**Review by ibreakthecloud — COMMENTED** [review]
sorry for nitpicks :)

**Review by ibreakthecloud — APPROVED** [review]

**Review by noboruma — APPROVED** [review]

**Pull request merged**